src/modules/landing_target_estimator/KFxyzCoupledMoving.cpp

```diff
@@ -0,0 +1,483 @@
+	_covariance(4, 9) = _covariance(9, 4);
+	_covariance(5, 9) = _covariance(9, 5);
+	// _covariance(6, 9) = _covariance(9, 6);
+	// _covariance(7, 9) = _covariance(9, 7);
+	// _covariance(8, 9) = _covariance(9, 8);
+
+	_covariance(0, 10) = _covariance(10, 0);
+	_covariance(1, 10) = _covariance(10, 1);
+	_covariance(2, 10) = _covariance(10, 2);
+	_covariance(3, 10) = _covariance(10, 3);
+	_covariance(4, 10) = _covariance(10, 4);
+	_covariance(5, 10) = _covariance(10, 5);
+	// _covariance(6, 10) = _covariance(0, 6);
+	// _covariance(7, 10) = _covariance(10, 7);
+	// _covariance(8, 10) = _covariance(10, 8);
+	// _covariance(9, 10) = _covariance(10, 9);
+
+	_covariance(0, 11) = _covariance(11, 0);
+	_covariance(1, 11) = _covariance(11, 1);
+	_covariance(2, 11) = _covariance(11, 2);
+	_covariance(3, 11) = _covariance(11, 3);
+	_covariance(4, 11) = _covariance(11, 4);
+	_covariance(5, 11) = _covariance(11, 5);
+	// _covariance(6, 11) = _covariance(11, 6);
+	// _covariance(7, 11) = _covariance(11, 7);
+	// _covariance(8, 11) = _covariance(11, 8);
+	// _covariance(9, 11) = _covariance(11, 9);
+	// _covariance(10, 11) = _covariance(11, 10);
+}
+
+
+bool KFxyzCoupledMoving::update()
+{
+	// Avoid zero-division
+	if (_innov_cov  <= 0.000001f && _innov_cov  >= -0.000001f) {
+		return false;
+	}
+
+	float beta = _innov / _innov_cov * _innov;
+
+	// 5% false alarm probability
+	if (beta > 3.84f) {
+		return false;
+	}
+
+	matrix::Matrix<float, 12, 1> kalmanGain = _covariance * _meas_matrix.transpose() / _innov_cov;
+
+	_state = _state + kalmanGain * _innov;
+	_covariance = _covariance - kalmanGain * _meas_matrix * _covariance;
+
+	return true;
+}
+
+void KFxyzCoupledMoving::setH(matrix::Vector<float, 12> h_meas)
+{
+	// h_meas = [rx, ry, rz, r_dotx, r_doty, r_dotz, bx, by, bz, atx, aty, atz]
+	_meas_matrix.row(0) = h_meas;
+}
+
+void KFxyzCoupledMoving::syncState(float dt, matrix::Vector<float, 3> acc)
+{
+	// Total ops: 45
+
+	const float _tmp0 = 0.5f * dt * dt;
+	const float _tmp1 = _state(3, 0);
+	const float _tmp2 = _state(4, 0);
+	const float _tmp3 = _state(5, 0);
+	const float _tmp4 = dt;
+	const float _tmp5 = _state(9, 0);
+	const float _tmp6 = _state(10, 0);
+	const float _tmp7 = _state(11, 0);
+
+	_sync_state(0, 0) = _state(0, 0) + _state(9, 0) * _tmp0 - _tmp0 * acc(0) - _tmp1 * dt;
+	_sync_state(1, 0) = _state(1, 0) + _state(10, 0) * _tmp0 - _tmp0 * acc(1) - _tmp2 * dt;
+	_sync_state(2, 0) = _state(11, 0) * _tmp0 +  _state(2, 0) - _tmp0 * acc(2) - _tmp3 * dt;
+	_sync_state(3, 0) = _tmp1 + _tmp4 * acc(0) - _tmp5 * dt;
+	_sync_state(4, 0) = _tmp2 + _tmp4 * acc(1) - _tmp6 * dt;
+	_sync_state(5, 0) = _tmp3 + _tmp4 * acc(2) - _tmp7 * dt;
+	_sync_state(6, 0) =  _state(6, 0);
+	_sync_state(7, 0) =  _state(7, 0);
+	_sync_state(8, 0) =  _state(8, 0);
+	_sync_state(9, 0) = _tmp5;
+	_sync_state(10, 0) = _tmp6;
+	_sync_state(11, 0) = _tmp7;
+}
+
+float KFxyzCoupledMoving::computeInnovCov(float meas_unc)
+{
+	// Total ops: 300
+
+	_innov_cov =
```

Comment:
Is that significantly cheaper than computing HPH'+R using matrix operations?

Reply:
Yes because only 300 operations are required to compute two 12x12 matrix multiplications (O(n^3)).